### PR TITLE
[Random] Random state management

### PR DESCRIPTION
**Pull request by comaniac**
## Description ##
- Add random state management to deal with the requirements of using the same or different random seeds in a TP group. The implementation is based on the one in Megatron-LM.
- Add an activation checkpointing with the consideration of random states.
- Add an op `DropoutWithTensorParallel` that can be replaced by users when writing a schedule.
- Add unit tests.
- Disable randomly plugin in pytest. This plugin makes our random seed setup in the test fixture useless.

Notes:
1. We now offer an API `set_random_seed` for users to call in the training script. Users have to manually call it and specify the rank of 3D parallelism.
2. All changes in this PR will have no impact if `set_random_seed` is not called in advance.
3. Fidelity testing shows the updated GPT schedule with 3D parallelism could align the loss to ZeRO-3 (with and without activation checkpointing), but flash attention has to be disabled.
4. I'll update the flash attention to the latest one and see if the problem will gone.

## Checklist ##

- [x] PR's title starts with a category (e.g. [Bugfix], [Model], [Tutorial], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

cc @szhengac @chhzh123

### CODE REVIEW

File: tests/test_shard_sync_op.py [hunk 1]
```diff
@@ -20,9 +20,6 @@
 
 def init_model_and_data(local_rank):
     model = nn.Linear(10, 10).cuda(local_rank)
-    # Make sure all devices have the same model.
-    dist.broadcast(model.weight.data, src=0)
-    dist.broadcast(model.bias.data, src=0)
```
Comment (chhzh123) [inline tests/test_shard_sync_op.py:25]: Why this is no longer needed?

File: conftest.py [hunk 2]
```diff
@@ -31,3 +31,17 @@ def destory_dist():
             pass
 
     request.addfinalizer(destory_dist)
+
+
+@pytest.fixture(scope="function", autouse=True)
+def random_seed():
```
Comment (chhzh123) [inline conftest.py:37]: Would `fix_random_seed` a better name?

**Review by chhzh123 — APPROVED** [review]
LGTM. Thanks @comaniac.

File: tests/test_shard_sync_op.py (hunk 1 above)
Comment (comaniac) [reply tests/test_shard_sync_op.py:25]: Because I disabled pytest randomly plugin.

File: conftest.py (hunk 2 above)
Comment (comaniac) [reply conftest.py:37]: This name doesn't matter as it is an autouse fixture.

File: examples/gpt/megatron_hf.py [hunk 3]
```diff
@@ -60,6 +60,10 @@ def get_model(
             delay_init=delay_init,
         )
         model, _ = slapo.build(sch, init_weights=model._init_weights)
+        # Note 1: We assume no DP and PP in this script.
+        # Note 2: This overrides Megatron random seed management, so we only use
+        #         this script for benchmarking.
+        slapo.set_random_seed(2013, None, None, sch.rank)
```
Comment (szhengac) [inline examples/gpt/megatron_hf.py:66]: if i understand correctly. all the DP ranks also use the same seed so the loss wouldn't be right, but we only use this script for benchmarking

**Comment by szhengac** [comment]
LGTM

**Review by szhengac — APPROVED** [review]
